### PR TITLE
Testing framework for install-cni template processing

### PR DESCRIPTION
**Pull request by marqc**
Adds tests to verify cni_spec generation by install-cni.sh script

### CODE REVIEW

**Comment by aojea** [comment]
This is impressive, let´s get a run on CI

**Comment by aojea** [comment]
@jingyuanliang why are the githb actions not running?

**Comment by jingyuanliang** [comment]
Add those new shell scripts to the target of shellcheck?

**Comment by aojea** [comment]
> Add those new shell scripts to the target of shellcheck?

those are tests, should we shelcheck the tests?

you are the experts here, it looks good to me but I´m very lazy with bash so you should review @jingyuanliang

**Comment by jingyuanliang** [comment]
> those are tests, should we shelcheck the tests?

Hmm I just had some nitpicks and wanted shellcheck to say it for me.

**Comment by jingyuanliang** [comment]
Okay I just tried it. shellcheck gives many useless findings. Don't bother to add it then.

**Comment by marqc** [comment]
I decided to add it to shellcheck. Even if findings are not important or false-positive, having the check in pipeline can prevent us from introducing bugs later on.

I also want to add a readme file with documentation on the testing framework (how it works, how to use it). I will update the PR soon.

**Comment by marqc** [comment]
@jingyuanliang please take a look, it's ready for review now

File: scripts/testcase/tpl-testcase.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+export ENABLE_BANDWIDTH_PLUGIN=false
+export ENABLE_CILIUM_PLUGIN=false
+export ENABLE_MASQUERADE=false
+export ENABLE_IPV6=false
+
+export CNI_SPEC_TEMPLATE='{}'
+
+# Method to be invoked before install-cni.sh call
+# You can configure required mocks here
+function before_test() {
```
Comment (jingyuanliang) [inline scripts/testcase/tpl-testcase.sh:18]: It looks like the pattern of url matching in `curl` and `route` mocking is everywhere. Add them to the template?

File: scripts/testdata/expected-directpath.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    {
+      "type": "ptp",
+      "mtu": 1460,
+      "ipam": {
+        "type": "host-local",
+        "ranges": [
+          [{"subnet": "10.52.1.0/24"}], [{"subnet": "2600:1900:4000:318:0:7:0:0/112"}]
+        ],
+        "routes": [
+          {"dst": "0.0.0.0/0"}, {"dst": "::/0"}
```
Comment (jingyuanliang) [inline scripts/testdata/expected-directpath.json:14]: This is not the expected directpath (meaning you didn't get the env right).

Directpath doesn't have a default IPv6 route. (only one to directpath prefix)

**Comment by jingyuanliang** [comment]
I can think of several other test cases, but we can start with the current ones.

File: scripts/test-install-cni.sh [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  # shellcheck disable=SC1090 source-path=testcase
+  . ./"$testcase"
+
+  # initializing testcase mocks
+  before_test
+
+  # running install-cni script
+  /bin/bash ./install-cni.sh >>test.log 2>>test.log
+  exit_code="$?"
+  if [ "0" != "${exit_code}" ] ; then #>/dev/null 2>&1
```
[the remark this answers is not in the document]
Comment (marqc) [reply scripts/test-install-cni.sh]: leftover from debugging, removed

File: scripts/testcase/testcase-basic.sh [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+export ENABLE_IPV6=false
+
+CNI_SPEC_TEMPLATE=$(cat testdata/spec-template.json)
+export CNI_SPEC_TEMPLATE
+
+function before_test() {
+
+  function curl() {
+    # shellcheck disable=SC2317
+    if [[ "$*" =~ http://metadata\.google\.internal/computeMetadata/v1/instance/network-interfaces/0 ]]; then
```
[the remark this answers is not in the document]
Comment (marqc) [reply scripts/testcase/testcase-basic.sh:17]: case statement does not support regexp matching used in these comparisons

File: scripts/testcase/tpl-testcase.sh (hunk 1 above)
Comment (marqc) [reply scripts/testcase/tpl-testcase.sh:18]: ok, added to template and moved route mock implementation to set of common mocks

File: scripts/testdata/expected-directpath.json (hunk 2 above)
Comment (marqc) [reply scripts/testdata/expected-directpath.json:14]: so it should use env `CNI_SPEC_IPV6_ROUTE` - right?

Comment (jingyuanliang) [reply scripts/testdata/expected-directpath.json]: right.

File: scripts/testcase/testcase-basic.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (jingyuanliang) [reply scripts/testcase/testcase-basic.sh:17]: Hmm but I don't see you using regexp in the patterns.

File: scripts/install-cni-testing.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+Testing install-cni.sh
+Running test [testcase/testcase-basic.sh]: PASS
+Running test [testcase/testcase-directpath.sh]: PASS
+Running test [testcase/testcase-dualstack.sh]: PASS
+Test execution log available in test.log
+```
+
+## Concept 
+
+Tests execute the `install_cni.sh` script with given set of env values (test
```
[the remark this answers is not in the document]
Comment (jingyuanliang) [reply scripts/install-cni-testing.md]: oops, sorry, .sh

File: scripts/testcase/testcase-basic.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (marqc) [reply scripts/testcase/testcase-basic.sh]: I was using it for partial match, but you are right, I can achieve the same results with shell pattern matching

File: scripts/testdata/expected-directpath.json (hunk 2 above)
Comment (marqc) [reply scripts/testdata/expected-directpath.json]: fixed

**Pull request merged**